From 8a9239513f03d892b6f768fa4f44e1d3facb0fab Mon Sep 17 00:00:00 2001
From: Sriram Sankar Sankarapandian Thangaraj <sriramsa@qti.qualcomm.com>
Date: Tue, 20 Jan 2026 12:18:08 +0530
Subject: [PATCH] UPSTREAM:CP:hostapd: Update wds_mld_uid for multiAP sta.

wds_mld_uid used for WDS STA AP\VLAN creation is was always 0 for a multiAP
STA causing issues in interface creation.

Fix added to get new wds_mld_uid before interface creation for multiAP STA
and set mld_uid, flags to all partner links.

Patch-dependency: I88c3a949035 (UPSTREAM:CP:hostapd: Set WDS flag for all link STA)

Signed-off-by: Sriram Sankar Sankarapandian Thangaraj <sriramsa@qti.qualcomm.com>
---
 src/ap/ieee802_11.c | 22 ++++++++++++++++------
 1 file changed, 16 insertions(+), 6 deletions(-)

diff --git a/src/ap/ieee802_11.c b/src/ap/ieee802_11.c
index 7e1641c..cbccd84 100644
--- a/src/ap/ieee802_11.c
+++ b/src/ap/ieee802_11.c
@@ -7239,7 +7239,7 @@ static void hostapd_ml_handle_assoc_cb(struct hostapd_data *hapd,
 #endif /* CONFIG_IEEE80211BE */
 }
 
-static void set_wds_sta_flag(struct hostapd_data *hapd, struct
+static void set_sta_flag_to_partner_links(struct hostapd_data *hapd, struct
 			     sta_info *sta)
 {
 	struct sta_info *psta;
@@ -7253,7 +7253,10 @@ static void set_wds_sta_flag(struct hostapd_data *hapd, struct
 			psta = ap_get_sta(phapd, sta->addr);
 			if (psta) {
 				psta->wds_mld_uid = aid;
-				psta->flags |= WLAN_STA_WDS;
+				psta->flags = ((psta->flags & ~WLAN_STA_WDS) |
+						(sta->flags & WLAN_STA_WDS));
+				psta->flags = ((psta->flags & ~WLAN_STA_MULTI_AP) |
+						(sta->flags & WLAN_STA_MULTI_AP));
 			}
 		}
 	}
@@ -7381,7 +7384,7 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 			   MAC2STR(sta->addr));
 		sta->pending_wds_enable = 0;
 		sta->flags |= WLAN_STA_WDS;
-		set_wds_sta_flag(hapd, sta);
+		set_sta_flag_to_partner_links(hapd, sta);
 		if (hapd->conf->mld_ap && hostapd_get_wds_mld_sta_uid(hapd, sta) < 0) {
 			wpa_printf(MSG_DEBUG, "No room for uid"
 				   "to enable 4-address WDS mode for STA "
@@ -7399,14 +7402,21 @@ static void handle_assoc_cb(struct hostapd_data *hapd,
 		char ifname_wds[IFNAMSIZ + 1];
 		int aid;
 
-		if (hapd->conf->mld_ap)
+		if (hapd->conf->mld_ap) {
+			if (hostapd_get_wds_mld_sta_uid(hapd, sta) < 0) {
+				wpa_printf(MSG_DEBUG, "No room for uid"
+						"to enable 4-address WDS mode for STA "
+						MACSTR, MAC2STR(sta->addr));
+				return;
+			}
 			aid = sta->wds_mld_uid;
-		else
+		} else
 			aid = sta->aid;
 
 		wpa_printf(MSG_DEBUG, "Reenable 4-address WDS mode for STA "
 			   MACSTR " (aid %u)",
 			   MAC2STR(sta->addr), aid);
+		set_sta_flag_to_partner_links(hapd, sta);
 		ret = hostapd_set_wds_sta(hapd, ifname_wds, sta->addr,
 					  aid, 1);
 		if (!ret)
@@ -7769,7 +7779,7 @@ void ieee802_11_rx_from_unknown(struct hostapd_data *hapd, const u8 *src,
 					"mode for STA with id %u flags 0x%x", aid, sta->flags);
 
 			sta->flags |= WLAN_STA_WDS;
-			set_wds_sta_flag(hapd, sta);
+			set_sta_flag_to_partner_links(hapd, sta);
 			ret = hostapd_set_wds_sta(hapd, ifname_wds,
 						  sta->addr, aid, 1);
 			if (!ret)
-- 
4.34.1

